File: Java/com/leetcode/spiralMatrixII/SpiralMatrixII.java
package com.leetcode.spiralMatrixII;

public class SpiralMatrixII {
	public int[][] generateMatrix(int n) {
        if (n == 0) {
            return new int[0][0];
        }
        
        int[][] matrix = new int[n][n];
        
        int counter = 0,
            count = n * n;
        
        int top = 0,
            down = n - 1,
            left = 0,
            right = n - 1,
            num = 0;
        boolean isFinished = false;
        while (true) {
            for (int i = left; i < n; i++) {
                if (num != count && matrix[top][i] == 0) {
                    num++;
                    matrix[top][i] = num;
                } else if (num == count){
                    isFinished = true;
                    break;
                }
            }
            
            if (isFinished || num == count) {
                break;
            }
            top++;
            
            for (int i = top; i < n; i++) {
                if (num != count && matrix[i][right] == 0) {
                    num++;
                    matrix[i][right] = num;
                    
                } else if (num == count) {
                    isFinished = true;
                    break;
                }
            }
            
            if (isFinished || num == count) {
                break;
            }
            right--;
            
            for (int i = right; i >= 0; i--) {
                if (num != count && matrix[down][i] == 0) {
                    num++;
                    matrix[down][i] = num;
                    
                } else if (num == count) {
                    isFinished = true;
                    break;
                }
            }
            
            if (isFinished || num == count) {
                break;
            }
            down--;
            
            for (int i = down; i >= 0; i--) {
                if (num != count && matrix[down][i] == 0) {
                    num++;
                    matrix[i][left] = num;
                    
                } else if (num == count) {
                    isFinished = true;
                    break;
                }
            }
            
            if (isFinished || num == count) {
                break;
            }
            
            left++;
            
        }
        
        return matrix;
    }
}

File: Python/spiral_matrix_ii/spiral_matrix_ii.py
# # pip install enum34 before python3.4

# from enum import Enum
# class D(Enum):
# 	right = 0
# 	down = 1
# 	left = 2
# 	up = 3

# class Solution(object):
# 	def __init__(self):
# 		self.matrix = []
# 		self.dir = D.right
# 		self.m = 0

# 	def generateMatrix(self, n):
# 		if n < 1:
# 			return []
# 		self.matrix = [[0 for x in range(n)] for x in range(n)]		
# 		self.m = len(self.matrix)
# 		self.fill(0, 0, 1)
# 		return self.matrix

# 	def fill(self, i, j, num):
# 		while i >= 0 and i < self.m and \
# 				j >= 0 and j < self.m and \
# 				self.matrix[i][j] == 0:
# 			self.matrix[i][j] = num
# 			if num == self.m * self.m:
# 				return
# 			if self.dir == D.right:
# 				j += 1
# 			elif self.dir == D.down:
# 				i += 1
# 			elif self.dir == D.left:
# 				j -= 1
# 			elif self.dir == D.up:
# 				i -= 1
# 			num += 1

# 		if self.dir == D.right:
# 			j -= 1
# 			i += 1
# 			self.dir = D.down
# 		elif self.dir == D.down:
# 			i -= 1
# 			j -= 1
# 			self.dir = D.left
# 		elif self.dir == D.left:
# 			j += 1
# 			i -= 1
# 			self.dir = D.up
# 		elif self.dir == D.up:
# 			i += 1
# 			j += 1
# 			self.dir = D.right

# 		self.fill(i, j, num)
	


class Solution(object):
	def __init__(self):
		self.matrix = []
		self.right = 0
		self.down = 1
		self.left = 2
		self.up = 3
		self.dir = self.right
		self.m = 0

	def generateMatrix(self, n):
		if n < 1:
			return []
		self.matrix = [[0 for x in range(n)] for x in range(n)]		
		self.m = len(self.matrix)
		self.fill(0, 0, 1)
		return self.matrix

	def fill(self, i, j, num):
		while i >= 0 and i < self.m and \
				j >= 0 and j < self.m and \
				self.matrix[i][j] == 0:
			self.matrix[i][j] = num
			if num == self.m * self.m:
				return
			if self.dir == self.right:
				j += 1
			elif self.dir == self.down:
				i += 1
			elif self.dir == self.left:
				j -= 1
			elif self.dir == self.up:
				i -= 1
			num += 1

		if self.dir == self.right:
			j -= 1
			i += 1
			self.dir = self.down
		elif self.dir == self.down:
			i -= 1
			j -= 1
			self.dir = self.left
		elif self.dir == self.left:
			j += 1
			i -= 1
			self.dir = self.up
		elif self.dir == self.up:
			i += 1
			j += 1
			self.dir = self.right

		self.fill(i, j, num)

